h2/src/main/org/h2/mvstore/MVStore.java

```diff
@@ -776,182 +795,127 @@ private void readStoreHeader() {
-                        Chunk c = Chunk.fromString(cursor.getValue());
-                        assert c.version <= currentVersion;
-                        // might be there already, due to meta traversal
-                        // see readPage() ... getChunkIfFound()
-                        Chunk test = chunks.putIfAbsent(c.id, c);
-                        if (test != null) {
-                            c = test;
-                        }
-                        assert chunks.get(c.id) == c;
-                        if ((test = validChunksByLocation.get(c.block)) != null && test.id == c.id) {
-                            continue;
-                        } else if ((test = validChunksById.get(c.id)) != null) {
-                            // We do not have a valid chunk at that location,
-                            // but there is a copy of same chunk from original
-                            // location.
-                            // Chunk header at original location does not have
-                            // any dynamic (occupancy) metadata, so it can't be
-                            // used here as is, re-point our chunk to original
-                            // location instead.
-                            c.block = test.block;
-                            continue;
-                        }
-
-                        if (!c.isLive()) {
-                            // we can just remove entry from meta, referencing to this chunk,
-                            // but store maybe R/O, and it's not properly started yet,
-                            // so lets make this chunk "dead" and taking no space,
-                            // and it will be automatically removed later.
-                            c.block = Long.MAX_VALUE;
-                            c.len = Integer.MAX_VALUE;
-                            if (c.unused == 0) {
-                                c.unused = creationTime;
-                            }
-                            if (c.unusedAtVersion == 0) {
-                                c.unusedAtVersion = INITIAL_VERSION;
+            vdiff = newest.version - headerVersion;
+            lastCandidates[(int)vdiff] = newest;
+            newest = test;
+        }
+        
+        // In most cases, the newest is the last chunk, otherwise the store is corrupted by
+        // H2 bug, a rogue thread or process, even re-order file system.
+        // It's unnecessary that verifying the all chunk validity in the meta of last chunk candidates:
+        // 1) If the store corrupted by a rogue thread or process and we fall back to the older chunk,
+        // then the data lost, it's very bad. We should suggest recovering the store by tool to user or 
+        // in recovery mode.
+        // 2) And if the newest chunk valid(both header and footer valid), the all chunks referenced 
+        // in the meta of the newest also valid, otherwise why is the newest valid?, or the store corrupted.
```

Comment:
I think, we discussed that assumption already. Due to possible out-of-order writes last chunk might be there, but one before might not. And yes,  "or store is corrupted" might very well be the case